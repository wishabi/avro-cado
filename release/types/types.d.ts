/// <reference types="node" />
import * as Avro from "avsc";
export interface Options {
    schemaRegistry: string;
    numRetries: number;
    schema: any;
    subject: string;
    wrapUnions?: string;
}
export interface EncodeFunc {
    (message: object): Buffer;
}
export declare type ResolverMap = {
    [index: number]: Promise<Avro.Resolver>;
};
export interface DecoderInfo {
    subject: string;
    schema: Avro.Type;
    resolversMap: ResolverMap;
    createSchemaResolver: Function;
}
export interface DecodeFunc {
    (buffer: Buffer): Promise<Object>;
}
