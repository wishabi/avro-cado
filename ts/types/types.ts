import * as Avro from "avsc";

/*
 ********************************************************
 *                                          PACKAGE TYPES
 ********************************************************
 */

export interface Options {
  schemaRegistry: string;
  numRetries: number;
  schema: any;
  subject: string;
  wrapUnions?: string;
}

/*
 ********************************************************
 *                                          ENCODER TYPES
 ********************************************************
 */

export interface EncodeFunc {
  (message: object): Buffer;
}

/*
 ********************************************************
 *                                          DECODER TYPES
 ********************************************************
 */

export type ResolverMap = { [index: number]: Promise<Avro.Resolver> };

export interface DecoderInfo {
  subject: string;
  schema: Avro.Type;
  resolversMap: ResolverMap;
  createSchemaResolver: Function;
}

export interface DecodeFunc {
  (buffer: Buffer): Promise<Object>;
}
