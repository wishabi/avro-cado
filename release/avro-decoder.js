"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Avro = require("avsc");
const rp = require("request-promise");
const config_1 = require("./config");
const util_1 = require("./util");
/**
 * Retrieve the Avro schema from the schema registry. On an error
 * eligible for retry, try up to the configure number of times
 *
 * @param id - the schema id that should be retrieved
 * @return - The Avro schema requested
 */
const genSchemaRetriever = ({ subject, schemaRegistry, numRetries }) => (id) => __awaiter(this, void 0, void 0, function* () {
    const req = {
        uri: `${schemaRegistry}/schemas/ids/${id}`,
        headers: [{ Accept: config_1.ACCEPT_HEADERS }],
        json: true,
        resolveWithFullResponse: true
    };
    let error = null;
    // implement retry on certain status/reason codes
    for (let i = 0; i <= numRetries; i += 1) {
        try {
            return JSON.parse((yield rp(req)).body.schema);
        }
        catch (err) {
            // save the error
            error = err;
            if (!util_1.handleError(err)) {
                break;
            }
        }
    }
    throw new Error(`Failed to retrieve schema for subject ${subject} with id ${id} :: ${error.message}`);
});
/**
 * Retrieve the Avro schema from the schema registry, parse it,
 * and create a resolver to the local Avro schema.
 *
 * @param subject - The subject under which the schema is registered
 * @param id - the schema id that should be the source for the
 *             schema resolver
 * @param schema - the target Avro schema for the resolver
 * @return - The Avro schema resolver used to convert messages
 *           encoded with the specified Avro schema to the local
 *           Avro schema format
 */
const genCreateSchemaResolver = ({ subject, retrieveSchema }) => (id, schema) => __awaiter(this, void 0, void 0, function* () {
    /**
     * Retrieve the schema by ID from the schema registry ...
     */
    const msgSchema = yield retrieveSchema(id);
    /**
     * ... parse the schema ...
     */
    const avSourceSchema = Avro.Type.forSchema(msgSchema);
    /**
     * ... and create a resolver to the schema we know how to consume
     */
    return schema.createResolver(avSourceSchema);
});
/**
 * Create and return an Avro schema resolver, if one doesn't already
 * exist for the schema id in the encoded message to the schema passed
 * in
 *
 * @param encoded - The Avro encoded buffer
 * @param schema - the target Avro schema for the resolver
 * @return - The promise that holds the Avro resolver
 */
const genGetSchemaResolver = ({ subject, resolversMap, createSchemaResolver }) => (encoded, schema) => __awaiter(this, void 0, void 0, function* () {
    // Do not attempt to get a resolver
    if (!schema) {
        return null;
    }
    if (!encoded) {
        return null;
    }
    /**
     * Extract the schema id with which the buffer was encoded
     */
    const id = encoded.slice(1, 5).readInt32BE(0);
    if (!resolversMap[id]) {
        /**
         * We do not have a resolver from the source schema to the
         * schema we support. Go and get a resolver.
         *
         * IMPORTANT
         *
         * We store a promise of a resolver in our hashmap. The reason is that
         * if we have multiple messages being handled asynchronously, we do not
         * want all of them to flood the schema registry with REST calls
         * to retrieve the schema
         */
        resolversMap[id] = createSchemaResolver(id, schema);
    }
    return resolversMap[id];
});
/**
 * Avro decode the passed in data
 *
 * @param encoded - The Avro encoded buffer
 * @param schema - the target Avro schema to which the decoded payload
 *                 should conform
 * @param resolver - a promise for an Avro resolver that will
 *                   convert the data from the schema with which
 *                   it was encoded to the schema which we want to
 *                   read
 * @return - The promise that holds the Avro decoded payload
 */
const decodePayload = (encoded, schema, resolver) => __awaiter(this, void 0, void 0, function* () {
    // Do not attempt to avro decode if we have no target schema specified
    if (!schema) {
        return encoded;
    }
    // Do not attempt to avro decode null value
    if (!encoded) {
        return encoded;
    }
    return schema.decode(encoded.slice(5), 0, yield resolver).value;
});
/**
 * Create an Avro decoder based on the specified parameters and
 * return a closure that takes an encoded Buffer and decodes it
 *
 * @param schema - the target Avro schema to which the decoded payload
 *                 should conform
 * @param getSchemaResolver - a function used to retrieve the resolver
 *                            object that will be used to decode the payload
 * @param buffer - The Avro encoded buffer
 * @return - The Avro decoded object that conforms to the specified schema
 */
const genMessageDecoder = ({ schema, getSchemaResolver }) => (buffer) => __awaiter(this, void 0, void 0, function* () {
    return yield decodePayload(buffer, schema, getSchemaResolver(buffer, schema));
});
/*****************************************************************/
/**                      EXPORTED INTERFACE                     **/
/*****************************************************************/
exports.createDecoder = (opts) => {
    // Aggregate the configuration values with defaults
    const mergedOpts = config_1.processOptions(opts);
    // decoder info
    const decoderInfo = {
        subject: mergedOpts.subject,
        schema: mergedOpts.schema ? Avro.Type.forSchema(mergedOpts.schema) : null,
        resolversMap: {},
        retrieveSchema: null,
        createSchemaResolver: null,
        getSchemaResolver: null
    };
    decoderInfo.retrieveSchema = genSchemaRetriever(mergedOpts);
    decoderInfo.createSchemaResolver = genCreateSchemaResolver(decoderInfo);
    decoderInfo.getSchemaResolver = genGetSchemaResolver(decoderInfo);
    return genMessageDecoder(decoderInfo);
};
//# sourceMappingURL=avro-decoder.js.map