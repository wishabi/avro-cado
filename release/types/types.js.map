{"version":3,"file":"types.js","sourceRoot":"","sources":["../../ts/types/types.ts"],"names":[],"mappings":""}