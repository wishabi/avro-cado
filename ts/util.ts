import { Options } from "./types/types";

export const RETRY_STATUS_CODE_500 = 500;
export const RETRY_ERROR_CODE_50002 = 50002;
export const RETRY_ERROR_CODE_50003 = 50003;

/**
 * Decide whether a schema registry error is eligible to
 * be retried
 *
 * @param err - the error in question
 * @return - true iff the error is eligible for retry
 */
export const handleError = (err: any): boolean => {
  if (
    err.statusCode &&
    err.statusCode === RETRY_STATUS_CODE_500 &&
    err.error &&
    err.error.error_code &&
    (err.error.error_code === RETRY_ERROR_CODE_50003 ||
      err.error.error_code === RETRY_ERROR_CODE_50002)
  ) {
    return true;
  }

  return false;
};

/**
 * Merge the user specified configuration options with the
 * default ones
 *
 * @param defaultConf - the default configuration options
 * @param overrideConf - the user specified configuration options
 * @return - merged configuration options object
 */
export const aggregateOptions = (
  defaultConf: Options,
  overrideConf: Options
): Options => {
  const aggOptions = Object.assign({}, defaultConf, overrideConf);

  if (!overrideConf) {
    return aggOptions;
  }

  // determine how to handle unions
  const validWrapOptions: Array<string> = ["always", "never", "auto"];
  if (
    validWrapOptions.filter(opt => opt === aggOptions.wrapUnions).length === 0
  ) {
    aggOptions.wrapUnions = defaultConf.wrapUnions;
  }

  return aggOptions;
};
