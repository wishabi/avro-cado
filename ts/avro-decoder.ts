import * as Avro from "avsc";
import * as rp from "request-promise";
import { ACCEPT_HEADERS, processOptions } from "./config";
import {
  DecodeFunc,
  DecoderInfo,
  Options,
  SchemaResolverFunc
} from "./types/types";
import { aggregateOptions, handleError } from "./util";

/**
 * Retrieve the Avro schema from the schema registry. On an error
 * eligible for retry, try up to the configure number of times
 *
 * @param subject - the subject under which the schema is registered
 * @param schemaRegistry - the schema registry path
 * @param numRetries - the number of retries
 * @param id - the schema id that should be retrieved
 * @return - The Avro schema requested
 */
export const retrieveSchema = async (
  { subject, schemaRegistry, numRetries }: Options,
  id: number
) => {
  const req = {
    uri: `${schemaRegistry}/schemas/ids/${id}`,
    headers: { Accept: ACCEPT_HEADERS },
    json: true,
    resolveWithFullResponse: true
  };

  let error = null;

  // implement retry on certain status/reason codes
  for (let i = 0; i <= numRetries; i += 1) {
    try {
      return JSON.parse((await rp(req)).body.schema);
    } catch (err) {
      // save the error
      error = err;

      if (!handleError(err)) {
        break;
      }
    }
  }

  throw new Error(
    `Failed to retrieve schema for subject ${subject} with id ${id} :: ${
      error.message
    }`
  );
};

/**
 * Retrieve the Avro schema from the schema registry, parse it,
 * and create a resolver to the local Avro schema.
 *
 * @param id - the schema id that should be the source for the
 *             schema resolver
 * @param schema - the target Avro schema for the resolver
 * @return - The Avro schema resolver used to convert messages
 *           encoded with the specified Avro schema to the local
 *           Avro schema format
 */
export const genCreateSchemaResolver = (
  opts: Options
): SchemaResolverFunc => async (
  id: number,
  schema: Avro.Type
): Promise<Avro.Resolver> => {
  /**
   * Retrieve the schema by ID from the schema registry ...
   */
  const msgSchema = await retrieveSchema(opts, id);

  /**
   * ... parse the schema ...
   */
  try {
    const avSourceSchema = Avro.Type.forSchema(msgSchema);

    /**
     * ... and create a resolver to the schema we know how to consume
     */
    return schema.createResolver(avSourceSchema);
  } catch (err) {
    throw new Error(
      `Local schema is not compatible with schema from registry with id ${id} :: ${
        err.message
      }`
    );
  }
};

/**
 * Create an Avro decoder based on the specified parameters and
 * return a closure that takes an encoded Buffer and decodes it
 *
 * @param schema - the target Avro schema to which the decoded payload
 *                 should conform
 * @param createSchemaResolver - a function used to create the resolver
 *                               object that will be used to decode the payload
 * @param resolversMap - a hashmap keys on the schema ids where the value is a
 *                       promise holding a schema resolver
 * @param buffer - The Avro encoded buffer
 * @return - The Avro decoded object that conforms to the specified schema
 */
export const genPayloadDecoder = ({
  schema,
  createSchemaResolver,
  resolversMap
}: DecoderInfo): DecodeFunc => async (
  buffer: Buffer | null
): Promise<object> => {
  // Do not attempt to avro decode null value
  if (!buffer) {
    return buffer;
  }

  /**
   * Extract the schema id with which the buffer was encoded
   */
  const id: number = buffer.slice(1, 5).readInt32BE(0);

  if (!resolversMap[id]) {
    /**
     * We do not have a resolver from the source schema to the
     * schema we support. Go and get a resolver.
     *
     * IMPORTANT
     *
     * We store a promise of a resolver in our hashmap. The reason is that
     * if we have multiple messages being handled asynchronously, we do not
     * want all of them to flood the schema registry with REST calls
     * to retrieve the schema
     */
    resolversMap[id] = createSchemaResolver(id, schema);
  }

  return schema.decode(buffer.slice(5), 0, await resolversMap[id]).value;
};

/*
 *****************************************************************
 *                      EXPORTED INTERFACE                       *
 *****************************************************************
 */

export const createDecoder = (opts: Options): DecodeFunc => {
  // Aggregate the configuration values with defaults
  const mergedOpts: Options = processOptions(opts);

  // decoder info
  const decoderInfo: DecoderInfo = {
    subject: mergedOpts.subject,
    schema: Avro.Type.forSchema(mergedOpts.schema),
    resolversMap: {},
    createSchemaResolver: null
  };

  decoderInfo.createSchemaResolver = genCreateSchemaResolver(mergedOpts);

  return genPayloadDecoder(decoderInfo);
};
