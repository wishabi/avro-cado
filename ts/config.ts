import { aggregateOptions } from "../ts/util";
import { IOptions } from "./types/types";

export const ACCEPT_HEADERS: string =
  "application/vnd.schemaregistry.v1+json, application/vnd.schemaregistry+json, application/json";

const optionsDefault: IOptions = {
  schemaRegistry: "http://localhost:8081",
  numRetries: 10, // number of attempts to call schemaRegistry
  wrapUnions: "auto", // avsc option
  subject: null, // subject for schema registration
  schema: null // schema object as needed for avsc
};

/**
 * Validate the options and merge them with the default values
 *
 * @param opts - the user specified options
 * @return - merged configuration options
 */
export const processOptions = (opts: IOptions): IOptions => {
  if (!opts.subject) {
    throw new Error("subject key missing");
  }

  if (!opts.schema) {
    throw new Error("schema key missing");
  }

  // Aggregate the configuration values with defaults
  return aggregateOptions(optionsDefault, opts);
};
