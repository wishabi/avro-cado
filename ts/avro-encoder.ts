import * as Avro from "avsc";
import axios, { AxiosResponse } from "axios";
import { ACCEPT_HEADERS, processOptions } from "./config";
import { EncodeFunc, Options } from "./types/types";
import { handleError } from "./util";

/**
 * Register the specified schema for the specified topic under the
 * specified subject.
 *
 * On an error eligible for retry, keep trying up to the specified number
 * of times. Once all retries have been exhausted, any error is considered
 * a fatal one.
 *
 *
 * @param subject - the subject under which to register the schema
 * @param schemaRegistry - the schema registry host
 * @param numRetries - the number of retries to register before throwing an error
 * @param schema - the schema to register
 *
 * @return - A Promise holding the id of the schema in the schema registry
 */
export const registerSchema = async ({
  subject,
  schemaRegistry,
  schema,
  numRetries,
}: Options): Promise<number> => {
  // craft the REST call to the schema registry
  const req = {
    method: "POST",
    url: `${schemaRegistry}/subjects/${subject}/versions`,
    headers: { Accept: ACCEPT_HEADERS },
    data: {
      schema: JSON.stringify(schema),
    },
    json: true,
  };

  let error = null;

  // implement retry on certain status/reason codes
  for (let i = 0; i <= numRetries; i += 1) {
    try {
      return ((await axios(req)) as AxiosResponse).data.id;
    } catch (err) {
      // save the error
      error = err;

      if (!handleError(err)) {
        break;
      }
    }
  }

  throw new Error(
    `Failed to register schema for subject ${subject} :: ${error.message}`,
  );
};

/**
 * Create a function that takes a message JSON object
 * and Avro encodes it
 *
 * @param payload - the JSON object to Avro encode
 * @return - The Avro encoded object as a Buffer
 */
export const genMessageEncoder = (
  schema: Avro.Type,
  schemaId: number,
): EncodeFunc => (payload: object): Buffer => {
  if (payload === null) {
    // no payload so return it
    return null;
  }

  // create the avro header
  const header = Buffer.alloc(5);

  // magic byte
  header.writeInt8(0, 0);

  // schema id
  header.writeInt32BE(schemaId, 1);

  // Avro encode the payload
  const buffer: Buffer = schema.toBuffer(payload);

  return Buffer.concat([header, buffer]);
};

/*****************************************************************/
/**                      EXPORTED INTERFACE                     **/
/*****************************************************************/

export const createEncoder = async (opts: Options): Promise<EncodeFunc> => {
  // Aggregate the configuration values with defaults
  const mergedOpts: Options = processOptions(opts);

  return genMessageEncoder(
    Avro.Type.forSchema(mergedOpts.schema, {
      wrapUnions: mergedOpts.wrapUnions,
    }),
    await registerSchema(mergedOpts),
  );
};
