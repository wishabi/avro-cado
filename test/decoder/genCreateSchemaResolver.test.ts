jest.mock("axios");

import * as Avro from "avsc";
import axios from "axios";
import { genCreateSchemaResolver } from "../../ts/avro-decoder";
import { Options } from "../../ts/types/types";

const AVRO_SCHEMA = {
  type: "record",
  name: "TestMessage",
  namespace: "com.flipp.node.kafka.TestMessage",
  doc: "Properties related to a TestMessage.",
  fields: [
    {
      name: "key",
      type: "string",
      doc: "The the key for the message",
    },
    {
      name: "text",
      type: "string",
      doc: "The text for the message",
    },
  ],
};

const AVRO_SCHEMA_INCOMPATIBLE = {
  type: "record",
  name: "TestMessage",
  namespace: "com.flipp.node.kafka.TestMessage2",
  doc: "Properties related to a TestMessage.",
  fields: [
    {
      name: "key",
      type: "string",
      doc: "The the key for the message",
    },
    {
      name: "text",
      type: "string",
      doc: "The text for the message",
    },
  ],
};

const AVRO_SCHEMA_OBJ = Avro.Type.forSchema(AVRO_SCHEMA);
const AVRO_SCHEMA_INCOMPATIBLE_OBJ = Avro.Type.forSchema(
  AVRO_SCHEMA_INCOMPATIBLE,
);
const SCHEMA_ID = 1;

const opts: Options = {
  subject: "subject",
  schemaRegistry: "host",
  numRetries: 1,
  schema: AVRO_SCHEMA,
};

const createSchemaResolver = genCreateSchemaResolver(opts);

describe("genCreateSchemaResolver", () => {
  it("should return a schema resolver on success", async () => {
    expect.assertions(1);

    axios.mockImplementationOnce(() => {
      return {
        data: {
          schema: JSON.stringify(AVRO_SCHEMA),
        },
      };
    });

    const schemaResolver = await createSchemaResolver(
      SCHEMA_ID,
      AVRO_SCHEMA_OBJ,
    );

    expect(schemaResolver).toMatchSnapshot();
  });

  it("should throw an exception on an incompatible schemas", async () => {
    expect.assertions(1);

    axios.mockImplementationOnce((params) => {
      return {
        data: {
          schema: JSON.stringify(AVRO_SCHEMA),
        },
      };
    });

    await expect(
      createSchemaResolver(SCHEMA_ID, AVRO_SCHEMA_INCOMPATIBLE_OBJ),
    ).rejects.toMatchSnapshot();
  });
});
