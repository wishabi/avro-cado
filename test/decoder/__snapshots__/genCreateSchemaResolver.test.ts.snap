// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genCreateSchemaResolver should return a schema resolver on success 1`] = `
Resolver {
  "_read": [Function],
  "_readerType": Object {
    "doc": "Properties related to a TestMessage.",
    "fields": Array [
      Object {
        "doc": "The the key for the message",
        "name": "key",
        "type": "string",
      },
      Object {
        "doc": "The text for the message",
        "name": "text",
        "type": "string",
      },
    ],
    "name": "com.flipp.node.kafka.TestMessage.TestMessage",
    "type": "record",
  },
  "itemsType": null,
  "size": 0,
  "symbols": null,
  "valuesType": null,
}
`;

exports[`genCreateSchemaResolver should throw an exception on an incompatible schemas 1`] = `[Error: Local schema is not compatible with schema from registry with id 1 :: no alias found for com.flipp.node.kafka.TestMessage.TestMessage]`;
